tendermint/src/block/commit_sig.rs

```diff
@@ -0,0 +1,78 @@
+//! CommitSig within Commit
+
+use crate::serializers;
+use crate::{account, Signature, Time};
+use serde::{de::Error as _, Deserialize, Deserializer, Serialize, Serializer};
+
+/// BlockIDFlag is used to indicate the validator has voted either for nil, a particular BlockID or was absent.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum BlockIDFlag {
+    /// BlockIDFlagAbsent - no vote was received from a validator.
+    BlockIDFlagAbsent = 1,
+    /// BlockIDFlagCommit - voted for the Commit.BlockID.
+    BlockIDFlagCommit = 2,
+    /// BlockIDFlagNil - voted for nil.
+    BlockIDFlagNil = 3,
+}
+
+impl BlockIDFlag {
+    /// Deserialize this type from a byte
+    pub fn from_u8(byte: u8) -> Option<BlockIDFlag> {
+        match byte {
+            1 => Some(BlockIDFlag::BlockIDFlagAbsent),
+            2 => Some(BlockIDFlag::BlockIDFlagCommit),
+            3 => Some(BlockIDFlag::BlockIDFlagNil),
+            _ => None,
+        }
+    }
+
+    /// Serialize this type as a byte
+    pub fn to_u8(self) -> u8 {
+        self as u8
+    }
+
+    /// Serialize this type as a 32-bit unsigned integer
+    pub fn to_u32(self) -> u32 {
+        self as u32
+    }
+}
+
+impl Serialize for BlockIDFlag {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        self.to_u8().serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for BlockIDFlag {
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        let byte = u8::deserialize(deserializer)?;
+        BlockIDFlag::from_u8(byte)
+            .ok_or_else(|| D::Error::custom(format!("invalid block ID flag: {}", byte)))
+    }
+}
+
+/// CommitSig represents a signature of a validator.
+/// It's a part of the Commit and can be used to reconstruct the vote set given the validator set.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+pub struct CommitSig {
+    /// Block ID FLag
+    pub block_id_flag: BlockIDFlag,
+
+    /// Validator address
+    #[serde(deserialize_with = "serializers::parse_non_empty_id")]
+    pub validator_address: Option<account::Id>,
+
+    /// Timestamp
+    pub timestamp: Time,
```

Comment:
I guess this would need an Option too ? I believe if we have BlockIDFlagAbsent, then all these fields are empty https://github.com/tendermint/tendermint/blob/master/types/block.go#L476. Otherwise, they should all be populated. Or do we not need it here because there's a sane default?


In general I think there are a number of cases where a set of fields are either all present or all absent, but if they each individually are encoded as Option, this information isn't clear, since it looks like they can all independently be present or absent. In these cases, maybe it's better to use enums with Present and Absent variants to better indicate this ... related https://github.com/informalsystems/tendermint-rs/issues/101 - but that's for a follow up PR

Reply:
Looks like the go-code sends a zero time. Meaning sec and ms are 0, or more accurately: `0001-01-01T00:00:00Z` is the JSON we'll see for a `BlockIDFlagAbsent` commit:
```go
csa := NewCommitSigAbsent()
b, _ := cdc.MarshalJSON(&csa)
fmt.Println(string(b))
```
yields:  `{"block_id_flag":1,"validator_address":"","timestamp":"0001-01-01T00:00:00Z","signature":null}`

Reply:
Opened issue #259 for this.